wled00/FX_fcn.cpp

```diff
@@ -1816,7 +1820,7 @@ bool WS2812FX::deserializeMap(uint8_t n) {
   return (customMappingSize > 0);
 }
 
-uint16_t IRAM_ATTR WS2812FX::getMappedPixelIndex(uint16_t index) const {
+__attribute__ ((always_inline)) inline uint16_t IRAM_ATTR WS2812FX::getMappedPixelIndex(uint16_t index) const {
```

Comment:
Non obligatory: I would prefer `__attribute__` at the end but `[[...]]` in front.

Reply:
I honestly have no Idea what the precompiler instructions mean in detail, I copied this from what they use in fastled... 
the idea behind this is to get rid of the 'function entry' instructions that are added when a function is called. When I added the `inline` flash size increased by a few bytes, telling me that it is actually inlined. Since this short function is only called from two places and is called A LOT this may be faster. I have no way to check (would need a proper debugger that shows assembly instructions being executed line by line).

Reply:
As I've recently learned these are compiler attributes.

Reply:
🤔 not sure if always_inline plays well with IRAM_ATTR .... the first tell the compiler to always inline the function, the latter says "put the function into IRAM" which means that a real function is needed.

Reply:
that is what I was wondering too, this is just a suggestion, i.e. to inline this for speed but how to tell the compiler to inline it to the functions that are in ram... not sure how it will do it.

Reply:
To my understanding:
* `inline` is a hint/suggestion to the compiler. So it might get inlined, or not.
* `__attribute__((always_inline))` is a directive. So the compiler _must_ inline this function, no matter if its efficient or not.

If you want to optimize function calls, its sometime useful to add `__attribute__((pure))` or `__attribute__((const))` to the function declaration. But only do this after double-checking that the code is actually "pure" (no side-effects) or "const" (solely depends on arguments). I did this in the MoonModules fork, but honestly it does not give you more than 1 or 2 fps even if you apply it to lots of functions.

https://github.com/MoonModules/WLED/commit/7f9da309c9d37a27226b0a2e247e47494fafa50b